EugeneChai/hr analytics.Rmd

```diff
@@ -0,0 +1,272 @@
+check_out_long<-check_out%>%pivot_longer(cols=all_dates,names_to="Dates",values_to="check_out_time")%>%mutate(id=str_c(X,Dates,sep=""))%>%select(id,check_out_time)
+
+```
+
+```{r}
+#Ensure that the check-in and check-out time variables are on the same dataframe and are converted into date objects. 
+check_in_out<-check_in_long%>%inner_join(check_out_long,by="id")%>%mutate(check_in_time=as.POSIXlt(check_in_time),check_out_time=as.POSIXlt(check_out_time))
+```
+
+```{r}
+#Create a new column that represents the average working hours for each employee.
+check_in_out<-check_in_out%>%mutate(working_hours=difftime(check_out_time,check_in_time,units="hours"))
+```
+
+```{r}
+check_in_out<-check_in_out%>%group_by(X)%>%summarise(avg_working_hours=mean(working_hours,na.rm=TRUE))%>%rename(EmployeeID=X)
+```
+
+Join all the relevant data together into one dataframe
+```{r}
+df<-general_data%>%inner_join(manager_survey_data,by="EmployeeID")%>%inner_join(employee_survey_data,by="EmployeeID")%>%inner_join(check_in_out,by="EmployeeID")
+```
+
+```{r}
+#Add a new column that determines if an employee has worked overtime or not
+df<-df%>%mutate(overtime=if_else(avg_working_hours>8,"Yes","No"))
+```
+
+
+# Data Imputation
+```{r}
+colSums(is.na(df))
+```
+```{r}
+#The null values in NumCompaniesWorked can be filled by using estimated values formed by using both YearsAtCompany and TotalWorkingYears.
+na_numcompaniesworked<-df %>% filter(is.na(NumCompaniesWorked))%>%select(EmployeeID,NumCompaniesWorked,YearsAtCompany,TotalWorkingYears)
+```
+
+```{r}
+#An important statistics to calculate is the average years an employee works at a company.
+df%>%ggplot(aes(x=YearsAtCompany))+geom_density()
+```
+Since the distribution is highly skewed to the right, it is best to use median to estimate the average number of years an employee works at a company. 
+```{r}
+df$NumCompaniesWorked[is.na(df$NumCompaniesWorked)]<-na_numcompaniesworked$TotalWorkingYears%/%median(df$YearsAtCompany,na.rm = TRUE)
+```
+
+```{r}
+# The mode is used to replace the null values for EnvironmentSatisfaction
+df%>%ggplot(aes(x=EnvironmentSatisfaction))+geom_bar()
+df$EnvironmentSatisfaction[is.na(df$EnvironmentSatisfaction)]<-3
+```
+
+```{r}
+# The mode is used to replace the null values for JobSatisfaction
+df%>%ggplot(aes(x=JobSatisfaction))+geom_bar()
+df$JobSatisfaction[is.na(df$JobSatisfaction)]<-4
+```
+
+```{r}
+# The mode is used to replace the null values for WorkLifeBalance
+df%>%ggplot(aes(x=WorkLifeBalance))+geom_bar()
+df$WorkLifeBalance[is.na(df$WorkLifeBalance)]<-3
+```
+
+```{r}
+#To replace null values for TotalWorkingYears,estimated values are calculated using NumCompaniesWorked and TotalWorkingYears
+df %>% filter(is.na(TotalWorkingYears))%>%select(EmployeeID,NumCompaniesWorked,YearsAtCompany,TotalWorkingYears)
+```
+```{r}
+#Manually fill in values where NumCompaniesWorked is equal to 0 or 1
+df[2368,20]<-10
+df[4410,20]<-21
+df[24,20]<-25
+df[3819,20]<-6

```

Comment:
Could this be done in a function? This is me being very nitpicky btw and I like to automate anything where possible 👍🏼